File: packages/whisker-video/android/src/main/kotlin/rs/whisker/elements/video/VideoModule.kt
// `whisker-video` ModuleDefinition (Android).
//
// The `VideoView` Lynx UI subclass this references lives in
// `VideoView.kt`. Same split on iOS (`VideoModule.swift` +
// `VideoView.swift`).

package rs.whisker.elements.video

import rs.whisker.runtime.Module
import rs.whisker.runtime.WhiskerModule
import rs.whisker.runtime.ModuleDefinition
import rs.whisker.runtime.WhiskerValue

/**
 * DSL-driven module. [WhiskerModule] is the registration signal:
 * the KSP processor finds every annotated module and emits the
 * registration block into `WhiskerVideoBehaviors.registerAll()`.
 */

@WhiskerModule
class VideoModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("Video")
        View("whisker-video:Video", VideoView::class.java) {
            Prop("src") { view: VideoView, value ->
                view.setSrc(value.asString() ?: "")
            }
            Function("play") { view: VideoView, _ -> view.play(); WhiskerValue.Null }
            Function("pause") { view: VideoView, _ -> view.pause(); WhiskerValue.Null }
            Function("seek") { view: VideoView, args ->
                view.seek(args.getOrNull(0)?.asDouble() ?: 0.0)
                WhiskerValue.Null
            }
        }
    }
}

File: packages/whisker-video/ios/Sources/WhiskerVideo/VideoModule.swift
// `whisker-video` ModuleDefinition (iOS).
//
// The `VideoView` Whisker module view this references lives in
// `VideoView.swift`. Same split on Android (`VideoModule.kt` +
// `VideoView.kt`).

import WhiskerModule    // Module, ModuleDefinition, DSL

/// DSL-driven module. `@WhiskerModule` is the registration signal:
/// the SwiftPM codegen plugin scans the target's sources for annotated
/// module declarations and emits a registration block in
/// `<Target>+Generated.swift`, which registers the view and module functions
/// with Whisker's Host registry.
@WhiskerModule
public final class VideoModule: Module {
    public override func definition() -> ModuleDefinition {
        ModuleDefinition {
            Name("Video")
            View("whisker-video:Video", VideoView.self) {
                Prop("src", clear: { (view: VideoView) in view.setSrc("") }) { (view: VideoView, value: WhiskerValue) in
                    view.setSrc(value.asString ?? "")
                }
                Command("play")  { (view: VideoView, _: WhiskerValue) in view.play() }
                Command("pause") { (view: VideoView, _: WhiskerValue) in view.pause() }
                Command("seek")  { (view: VideoView, parameters: WhiskerValue) in
                    view.seek(parameters.asDouble ?? 0)
                }
            }
        }
    }
}
